src/main/java/org/gradle/java/JigsawPlugin.java

```diff
@@ -92,20 +203,94 @@ public void execute(Task task) {
                     @Override
                     public void execute(CopySpec copySpec) {
                         copySpec.from(moduleDir);
-                        copySpec.into(main.getAllJava().getOutputDir());
+                        copySpec.into(main.getJava().getOutputDir());
                     }
                 });
             }
         });
     }
 
+    private void configureCompileTestJavaTask(final Project project) {
+        final JavaCompile compileTestJava = getCompileTestJavaTask(project);
+        final SourceSet main = ((SourceSetContainer) project.getProperties().get("sourceSets")).getByName("main");
+        final SourceSet test = ((SourceSetContainer) project.getProperties().get("sourceSets")).getByName("test");
+        final JigsawModule module = (JigsawModule) project.getExtensions().getByName(EXTENSION_NAME);
+        compileTestJava.doFirst(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                compileTestJava.setClasspath(project.files());
+                List<String> args = new ArrayList<>();
+                args.add("--module-path");
+                args.add(project.files(
+                        main.getJava().getOutputDir(),
+                        project.getConfigurations().getByName("testCompileClasspath")).getAsPath());
+                args.add("--add-modules");
+                args.add("junit");
+                args.add("--add-reads");
+                args.add(String.format("%s=junit", module.getName()));
+                args.add("--patch-module");
+                args.add(String.format("%s=%s", module.getName(), test.getJava().getSourceDirectories().getAsPath()));
+                compileTestJava.getOptions().setCompilerArgs(args);
+            }
+        });
+    }
+
+    private JavaExec getRunTask(Project project) {
+        Set<JavaExec> tasks = project.getTasks().withType(JavaExec.class);
```

Comment:
Again `tasks.findByName(..)` is probably enough. If you want to validate the type, you can still do it with an `instanceof` check.